### PR TITLE
Team2sprint2

### PR DESCRIPTION
**Pull request by travela**
### Sprint 1

We adapted all requested changes from the review and merged the latest version of the README into this pull request to make sure there won't be any merge conflicts. The previous PR will be removed.

### Sprint 2

We implemented all basic functions to perform a PIA (Property Inference Attack).

### CODE REVIEW

**Comment by fraboeni** [comment]
Well done!

File: privacy_evaluator/attacks/attack_interface.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self, model, training_data, test_data):
+        self.model = model
+        self.training_data = training_data
+        self.test_data = test_data
+
+    """ Performs the actual attack on the model
+    :param prams: The prarameters of the Attack
+    :type params: dict
+    return: result of the attack
+    rtype: TODO: what Type?
```
Comment (fraboeni) [inline privacy_evaluator/attacks/attack_interface.py:24]: Please never have TODOs in a final version. I merged it now because there are no mayor issues, but please change that.

File: privacy_evaluator/attacks/property_inference_attack.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    ]:
+        """
+        Create the shadow training sets, half fulfill the property, half fulfill the negation of the property.
+        The function works for the specific binary case that the property is a fixed distribution specified in the input
+        and the negation of the property is a balanced distribution.
+        :param dataset: Dataset out of which shadow training sets should be created
+        :param amount_sets: how many shadow training sets should be created
+        :param size_set: size of one shadow training set for one shadow classifier
+        :param property_num_elements_per_classes: number of elements per class, this is the property
+        :return: shadow training sets for property,
```
Comment (fraboeni) [inline privacy_evaluator/attacks/property_inference_attack.py:44]: Please add also the types INSIDE the dictionary.

File: privacy_evaluator/metrics/accuracy.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+def accuracy_difference(train_accuracy, test_accuracy):
+    """Calculate the difference between the train and test accuracy.
+
+    The difference is calculated by subtracting the test accuracy from thee train accuracy. The result is a positive
```
Comment (fraboeni) [inline privacy_evaluator/metrics/accuracy.py:4]: There is still the typo that we have already talked about several times, so please change it.

File: privacy_evaluator/metrics/accuracy.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    :return: The difference between the train and test accuracy.
+    :rtype: int, float
+    """
+    return train_accuracy - test_accuracy
+
+
+def accuracy_ratio(train_accuracy, test_accuracy):
+    """Calculate the ratio between the train and test accuracy.
+
+    The ratio is calculated by dividing the test accuracy from thee train accuracy. The result is higher than 1 if the
```
Comment (fraboeni) [inline privacy_evaluator/metrics/accuracy.py:22]: Also this text should be changed.

**Review by fraboeni — COMMENTED** [review]
Thank you for including these small changes.

File: privacy_evaluator/attacks/attack_interface.py (hunk 1 above)
Comment (InaFen) [reply privacy_evaluator/attacks/attack_interface.py:24]: This is the attack interface from team 1. I added it as a subject for tomorrow.

File: privacy_evaluator/metrics/accuracy.py (hunk 3 above)
Comment (InaFen) [reply privacy_evaluator/metrics/accuracy.py:4]: This is the accuracy function from team 1. I added it as a subject for tomorrow.

File: privacy_evaluator/metrics/accuracy.py (hunk 4 above)
Comment (InaFen) [reply privacy_evaluator/metrics/accuracy.py:22]: This is the accuracy function from team 1. I added it as a subject for tomorrow.

File: privacy_evaluator/attacks/attack_interface.py (hunk 1 above)
Comment (InaFen) [reply privacy_evaluator/attacks/attack_interface.py:24]: @marisanest @jtorhoff @Erdnaf @budmil @blauertee Please adapt this according to Franziska's comment.

File: privacy_evaluator/metrics/accuracy.py (hunk 3 above)
Comment (InaFen) [reply privacy_evaluator/metrics/accuracy.py:4]: @marisanest @jtorhoff @Erdnaf @budmil @blauertee Please adapt this according to Franziska's comment.

File: privacy_evaluator/metrics/accuracy.py (hunk 4 above)
Comment (InaFen) [reply privacy_evaluator/metrics/accuracy.py:22]: @marisanest @jtorhoff @Erdnaf @budmil @blauertee Please adapt this according to Franziska's comment.

File: privacy_evaluator/attacks/property_inference_attack.py (hunk 2 above)
Comment (InaFen) [reply privacy_evaluator/attacks/property_inference_attack.py:44]: We discussed this in the sprint review, the types inside the dictionary were already given.

**Comment by jtorhoff** [comment]
None of the issues that you tagged us in are relevant anymore. All of them have been already addressed. You can check the state in the `team1sprint4` branch. It would also be a very strange workflow for us to commit changes to your branch, so I'm not sure what's going on here?

**Comment by InaFen** [comment]
@jtorhoff The background to the requests is that Franziska asked us to adapt some of your team's code that she thought was ours. We talked about it in the sprint review and decided that it would be best if your team would change it. Since you already did, there is nothing left for you to do and I will remove the "team1needed" tag.
